### PR TITLE
[Core] Rework ldap authentication

### LINKED ISSUE #1081: [Core] Configurable LDAP Information
**Is your feature request related to a problem? Please describe.**
When deploying the Feedbackssystem in different environments it is nesecarry to adjust the used ldap server. Currently the LDAP-Server-Informations are harde coded in [`application.yml`](https://github.com/thm-mni-ii/feedbacksystem/blob/e00b772a7f7751cc764957cd2d0cf29c4d271c6c/modules/fbs-core/api/src/main/resources/application.yml).

**Describe the solution you'd like**
Just like other configuration options the ldap configuration should be set able via environment variables for example `LDAP_BASEDN` and `LDAP_URL`.

### PR DESCRIPTION
**Pull request by Zitrone44**
To test be in the THM VPN and add the following to the `docker-compose.yml` environment section of core.

```yaml
- LDAP_ENABLED=true
```
---
closes #1081

### CODE REVIEW

File: modules/fbs-core/api/src/main/scala/de/thm/ii/fbs/controller/LoginController.scala [hunk 1]
```diff
@@ -124,13 +122,20 @@ class LoginController extends CasClientConfigurerAdapter {
...
+    } yield (username, password)
 
-    login match {
+    val user = credentials.flatMap(creds =>
+        userService.find(creds._1, creds._2).orElse(for {
+            ldapLogin <- ldapService.login(creds._1, creds._2)
+            ldapUser <- loadUserFromLdap(ldapLogin.getAttribute("uid").getStringValue)
+              .map(user => userService.find(user.username).getOrElse(userService.create(user, null)))
+          } yield ldapUser)
+    )
```
Comment (mxsph) [inline modules/fbs-core/api/src/main/scala/de/thm/ii/fbs/controller/LoginController.scala:136]: I would not put the ldap login in the route `.../api/v1/login/local` but make a new route `.../api/v1/login/` which tries several logins.

File: modules/fbs-core/api/src/main/scala/de/thm/ii/fbs/services/security/LdapService.scala [hunk 2]
```diff
@@ -0,0 +1,86 @@
...
+    val dnResolver = SearchDnResolver.builder
+      .factory(connectionFactory)
+      .dn(baseDn)
+      .filter(ldapFilter)
+      .subtreeSearch(true)
+      .build
+
+    val authHandler = new SimpleBindAuthenticationHandler(connectionFactory)
+    val auth = new Authenticator(dnResolver, authHandler)
+    val response = auth.authenticate(new AuthenticationRequest(username, new Credential(password), "uid"))
```
Comment (mxsph) [inline modules/fbs-core/api/src/main/scala/de/thm/ii/fbs/services/security/LdapService.scala:39]: `"uid"` should be changeable by a configuration value

File: modules/fbs-core/api/src/main/scala/de/thm/ii/fbs/services/security/LdapService.scala [hunk 3]
```diff
@@ -0,0 +1,86 @@
...
+  def getEntryByUid(uid: String): Option[LdapEntry] = runWithConnectionFactory(connectionFactory => {
+    val search = new SearchOperation(connectionFactory, baseDn)
+    val result = search.execute(
+      FilterTemplate.builder()
+        .filter(ldapFilter)
+        .parameter("user", uid)
+        .build(),
+      "uid", "sn", "givenName", "mail"
+    )
+    Option(result.getEntry)
```
Comment (mxsph) [inline modules/fbs-core/api/src/main/scala/de/thm/ii/fbs/services/security/LdapService.scala:56]: The following options are missing, but were available before the change

```
executor.setTimeLimit(Duration.ofSeconds(TIME_OUT))
executor.setDerefAliases(DerefAliases.ALWAYS)
```

**Review by mxsph — CHANGES_REQUESTED** [review]

**Review by mxsph — CHANGES_REQUESTED** [review]
- [x] The New login Route should be added to the Brute force protection

   https://github.com/thm-mni-ii/feedbacksystem/blob/c3671d1065625d2283b8026908fb18cd52a23b03/ws/src/main/scala/de/thm/ii/fbs/security/AntiBruteForceInterceptor.scala#L23

**Review by mxsph — DISMISSED** [review]

**Review by Medienkunzt — COMMENTED** [review]
**BugReport**
Setup test enviroment:
Delete all Docker Containers, Volumes and Images
Add LDAP_ENABLED=true
Run docker-compose up -d --build
Connect to VPN
Enter https://localhost/login

Try to login with my Username and Password
Result -> Internal Server Error (see Logs)

Header:
Request URL: https://localhost/api/v1/login/unified
Request Method: POST
Status Code: 500 
Remote Address: 127.0.0.1:443
Referrer Policy: strict-origin-when-cross-origin

Payload: {
  "username": "jk****",
  "password": "*****************"
}

Response: {"timestamp":1674474697786,"status":500,"error":"Internal Server Error","path":"/api/v1/login/unified"}

[logs.txt](https://github.com/thm-mni-ii/feedbacksystem/files/10479502/logs.txt)

**Review by Medienkunzt — APPROVED** [review]

**Review by mxsph — APPROVED** [review]

**Pull request merged**